File: src/components/ai/ann/ffi/c/ann.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "include/genann/genann.c"

int TrainAndPredictOnce(const double** input, const double** output, const int sample_total, const double* test_input, double* test_output, const int inputs, const int hidden_layers, const int hidden, const int outputs, const double learning_rate, const int epochs) {
    srand(time(0));

    genann *ann = genann_init(inputs, hidden_layers, hidden, outputs);

    int i, j;
    for (i = 0; i < epochs; ++i) {
        for (j = 0; j < sample_total; ++j) {
            genann_train(ann, *(input + j), *(output + j), learning_rate);
        }
    }

    *test_output = *genann_run(ann, test_input);

    genann_free(ann);

    return 0;
}

//demo
//int main(int argc, char *argv[])
//{
//    int i;
//
//    const double input[4][2] = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
//    const double* inputPointer[4];
//    for (i = 0; i < 4; ++i) {
//        inputPointer[i] = &(input[i]);
//    }
//
//    const double output[4][1] = {{0}, {1}, {1}, {0}};
//    const double* outputPointer[4];
//    for (i = 0; i < 4; ++i) {
//        outputPointer[i] = &(output[i]);
//    }
//
//    const double test_input[2] = {0, 1};
//    double test_output[1];
//
//    TrainAndPredictOnce(inputPointer, outputPointer, 4, test_input, test_output, 2, 1, 2, 1, 3, 500);
//
//    printf("%.1f", test_output[0]);
//
//    return 0;
//}

File: src/components/ai/ann/ffi/Ann.php
<?php

namespace SwFwLess\components\ai\ann\ffi;

use SwFwLess\components\utils\OS;
use SwFwLess\components\utils\runtime\PHPRuntime;

class Ann
{
    protected $config = [];
    protected $ffiPath;
    protected $udf;

    /**
     * @param array $config
     * @return Ann|static
     */
    public static function create($config = [])
    {
        return new self($config);
    }

    public function __construct($config = [])
    {
        $this->config = $config;

        if (PHPRuntime::supportFFI()) {
            $osType = OS::type();
            if ($osType === OS::OS_LINUX) {
                $this->ffiPath = __DIR__ . '/c/build/linux/libcann.so';
            } elseif ($osType === OS::OS_DARWIN) {
                $this->ffiPath = __DIR__ . '/c/build/darwin/libcann.so';
            }
        }

        if ($this->ffiPath) {
            $this->udf = $this->createUdf($this->ffiPath);
        }
    }

    public function createCDoubleNumbers($count)
    {
        return \FFI::new('double[' . ((string)$count) . ']');
    }

    public function createCConstDoubleNumbers($count)
    {
        return \FFI::new('const double[' . ((string)$count) . ']');
    }

    public function createCDoubleNumbersPointerArray($count)
    {
        return \FFI::new('double* [' . ((string)$count) . ']');
    }

    protected function createUdf($ffiPath)
    {
        return \FFI::cdef(
            "int TrainAndPredictOnce(const double** input, const double** output, const int sample_total, const double* test_input, double* test_output, const int inputs, const int hidden_layers, const int hidden, const int outputs, const double learning_rate, const int epochs);",
            $ffiPath
        );
    }

    public function trainAndPredictOnce(
        $input, $output, $sampleTotal, $testInput, $inputs, $hiddenLayers, $hidden,
        $outputs, $learningRate, $epochs
    )
    {
        $cInput = $this->createCDoubleNumbersPointerArray($sampleTotal);
        foreach ($input as $row => $numbers) {
            $cInputNumbersVarName = 'cInputNumbers' . $row;
            $$cInputNumbersVarName = $this->createCDoubleNumbers($inputs);
            foreach ($numbers as $col => $number) {
                $$cInputNumbersVarName[$col] = $number;
            }
            $cInput[$row] = \FFI::cast('double*', \FFI::addr($$cInputNumbersVarName));
        }

        $cOutput = $this->createCDoubleNumbersPointerArray($sampleTotal);
        foreach ($output as $row => $numbers) {
            $cOutputNumbersVarName = 'cOutputNumbers' . $row;
            $$cOutputNumbersVarName = $this->createCDoubleNumbers($outputs);
            foreach ($numbers as $col => $number) {
                $$cOutputNumbersVarName[$col] = $number;
            }
            $cOutput[$row] = \FFI::cast('double*', \FFI::addr($$cOutputNumbersVarName));
        }

        $cTestInput = $this->createCDoubleNumbers($inputs);
        foreach ($testInput as $i => $number) {
            $cTestInput[$i] = $number;
        }

        $cTestOutput = $this->createCDoubleNumbers($outputs);

        $this->udf->TrainAndPredictOnce(
            $cInput, $cOutput, $sampleTotal, $cTestInput,
            $cTestOutput, $inputs, $hiddenLayers, $hidden,
            $outputs, $learningRate, $epochs
        );

        return $cTestOutput;
    }
}
